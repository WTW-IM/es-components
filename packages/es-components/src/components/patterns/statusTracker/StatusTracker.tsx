import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import genName from '../../util/generateAlphaName';

const List = styled.ol`
  display: flex;
  margin: 0;
  padding: 0;
  position: relative;
  z-index: 1;
`;

const ListItem = styled.li<{ itemWidth: number }>`
  color: ${props => props.theme.colors.gray6};
  list-style-type: none;
  position: relative;
  text-align: center;
  width: ${props => props.itemWidth}%;

  .status-text {
    display: none;
  }

  &.done > .status-text {
    color: ${props => props.theme.brandColors.primary1};
  }

  &.active > .status-text,
  &.error > .status-text {
    color: ${props => props.theme.brandColors.primary1};
    display: inherit;
    font-weight: bold;
  }

  &:before {
    background-color: white;
    border: 3px solid ${props => props.theme.colors.gray5};
    border-radius: 50%;
    content: '';
    display: block;
    height: 20px;
    line-height: 20px;
    margin: 5px auto 15px auto;
    text-align: center;
    width: 20px;
  }

  &:after {
    background-color: ${props => props.theme.colors.gray5};
    content: '';
    height: 4px;
    left: -50%;
    position: absolute;
    top: 17px;
    width: 100%;
    z-index: -1;
  }

  &:first-child:after {
    content: none;
  }

  &.done:before {
    border-color: ${props => props.theme.brandColors.primary1};
    color: ${props => props.theme.brandColors.primary1};
    content: '\\e67b';
    font-family: 'bds-func-icons';
    font-size: 14px;
    font-weight: bold;
  }

  &.active:before {
    border-color: ${props => props.theme.brandColors.primary1};
    color: ${props => props.theme.brandColors.primary1};
    content: '\\e67b';
    font-family: 'bds-func-icons';
    font-size: 22px;
    font-weight: bold;
    height: 30px;
    line-height: 30px;
    margin: 0 auto 10px auto;
    width: 30px;
  }

  &.error:before {
    border-color: ${props => props.theme.colors.danger};
    color: ${props => props.theme.colors.danger};
    content: '\\e68e';
    font-family: 'bds-func-icons';
    font-size: 18px;
    font-weight: bold;
    height: 30px;
    line-height: 30px;
    margin: 0 auto 10px auto;
    width: 30px;
  }

  &.active:after,
  &.done:after,
  &.error:after {
    background-color: ${props => props.theme.brandColors.primary1};
  }

  @media (min-width: ${props => props.theme.screenSize.tablet}) {
    .status-text {
      display: inherit;
    }

    &.active:before,
    &.done:before {
      content: '\\e67b';
      font-family: 'bds-func-icons';
      font-size: 22px;
      height: 30px;
      line-height: 30px;
      margin: 0 auto 10px auto;
      width: 30px;
    }
  }
`;

export type StatusTrackerProps = Override<
  JSXElementProps<'ol'>,
  {
    statusArray: React.ReactNode[];
    step?: number;
    isErrorState?: boolean;
  }
>;

const StatusTracker = React.forwardRef<HTMLOListElement, StatusTrackerProps>(
  function ForwardedStatusTracker(
    { statusArray, step = 1, isErrorState, ...rest },
    ref
  ) {
    const itemWidth = 100 / statusArray.length;

    return (
      <List ref={ref} {...rest}>
        {statusArray.map((status, i) => {
          let statusClass;
          if (i + 1 === step) statusClass = isErrorState ? 'error' : 'active';
          if (i + 1 < step) statusClass = 'done';
          const itemKey = genName();

          return (
            <ListItem
              key={itemKey}
              itemWidth={itemWidth}
              className={statusClass}
            >
              <span className="status-text">{status}</span>
            </ListItem>
          );
        })}
      </List>
    );
  }
);

StatusTracker.propTypes = {
  /** Array of status descriptions which display sequentially under each step */
  statusArray: PropTypes.arrayOf(PropTypes.node).isRequired,
  /** Used to set the current active step (1-based) */
  step: PropTypes.number,
  /** Displays the active status with an error state instead of a check mark */
  isErrorState: PropTypes.bool
};

StatusTracker.defaultProps = {
  step: undefined,
  isErrorState: false
};

export default StatusTracker;
