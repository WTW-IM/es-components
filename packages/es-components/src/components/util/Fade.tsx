import React from 'react';
import PropTypes from 'prop-types';
import Transition, {
  TransitionStatus,
  TimeoutProps
} from 'react-transition-group/Transition';

type Opacity = React.CSSProperties['opacity'];

interface FadeProps extends Partial<TimeoutProps<HTMLElement>> {
  duration?: number;
  opacity?: Opacity;
  children: NonNullable<React.ReactNode>;
}

const getFadeStyle = (status: TransitionStatus, finalOpacity: Opacity) => {
  switch (status) {
    case 'entering':
    case 'entered':
      return { finalOpacity };
    default:
      return { opacity: 0 };
  }
};

type AnyReactHTMLElement = React.ReactHTMLElement<HTMLElement>;

const TransitionChild: React.FC<{
  children: AnyReactHTMLElement;
  style: React.CSSProperties;
  transitionStatus: TransitionStatus;
}> = ({ children, style, transitionStatus }) => {
  return React.cloneElement(children, {
    style: {
      ...((children?.props?.style as Maybe<React.CSSProperties>) || {}),
      ...style,
      ...getFadeStyle(
        transitionStatus,
        typeof style.opacity == 'number' || Boolean(style.opacity)
          ? style.opacity
          : 1
      )
    }
  });
};

const Fade: React.FC<FadeProps> = ({
  children,
  duration,
  opacity,
  ...otherProps
}) => {
  const transitionStyles = {
    transition: `opacity ${duration || 150}ms linear`,
    opacity
  };

  const transitionProps = {
    ...otherProps,
    timeout: duration || otherProps.timeout || 150
  };

  const hasValidChildren =
    children &&
    (typeof children === 'object' || typeof children === 'function');

  return !hasValidChildren ? (
    <>{children}</>
  ) : (
    <Transition {...transitionProps}>
      {transitionStatus => {
        try {
          return (
            React.Children.only(children) && (
              <TransitionChild
                transitionStatus={transitionStatus}
                style={transitionStyles}
              >
                {children as AnyReactHTMLElement}
              </TransitionChild>
            )
          );
        } catch (e) {
          return React.Children.map(children, child => {
            return !child ? (
              child
            ) : (
              <TransitionChild
                transitionStatus={transitionStatus}
                style={transitionStyles}
              >
                {child as AnyReactHTMLElement}
              </TransitionChild>
            );
          });
        }
      }}
    </Transition>
  );
};

Fade.propTypes = {
  children: PropTypes.any.isRequired,
  duration: PropTypes.number,
  opacity: PropTypes.number
};

Fade.defaultProps = {
  duration: 150,
  opacity: 1
};

export default Fade;
