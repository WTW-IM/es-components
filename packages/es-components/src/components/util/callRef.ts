import React from 'react';

const callRef = <T, R extends React.ForwardedRef<T>>(ref: Maybe<R>, value: T) =>
  typeof ref === 'function'
    ? ref(value)
    : ref && Object.hasOwn(ref, 'current') && (ref.current = value);

export default callRef;

export function callRefs<T, R extends React.ForwardedRef<T>>(
  value: T,
  ...refs: R[]
) {
  refs.forEach(r => callRef(r, value));
}
