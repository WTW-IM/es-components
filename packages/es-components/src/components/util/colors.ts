import tinycolor, { ColorInput } from 'tinycolor2';

export function darken(color: ColorInput, percent: number) {
  return tinycolor(color).darken(percent).toHexString();
}

export function lighten(color: ColorInput, percent: number) {
  return tinycolor(color).lighten(percent).toHexString();
}

export function getTextColor(
  bgColor: ColorInput,
  darkestColor: ColorInput = 'black',
  lightestColor: ColorInput = 'white'
) {
  return tinycolor.readability(darkestColor, bgColor) > 10
    ? darkestColor
    : lightestColor;
}
