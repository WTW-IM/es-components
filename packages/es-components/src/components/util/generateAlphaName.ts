export default function generateAlphaName(): string {
  const possible = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz';
  let text = '';
  for (let i = 0; i < 6; i++) {
    text += possible.charAt(Math.floor(Math.random() * possible.length));
  }
  return text;
}
