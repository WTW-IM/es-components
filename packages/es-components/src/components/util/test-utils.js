// eslint-env jest
import React from 'react';
import PropTypes from 'prop-types';
import { ThemeProvider } from 'styled-components';
import viaTheme from 'es-components-via-theme';
import { render } from '@testing-library/react';

export function ThemeComponent(props) {
  return <ThemeProvider theme={viaTheme} {...props} />;
}

ThemeComponent.propTypes = {
  children: PropTypes.node
};

export function renderWithTheme(component) {
  return render(<ThemeComponent>{component}</ThemeComponent>);
}
