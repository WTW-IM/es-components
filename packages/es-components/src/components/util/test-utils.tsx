import React, { Profiler } from 'react';
import PropTypes from 'prop-types';
import { ThemeProvider } from 'styled-components';
import viaTheme from 'es-components-via-theme';
import { render, RenderResult } from '@testing-library/react';

export function ThemeComponent(
  props: Omit<React.ComponentPropsWithoutRef<typeof ThemeProvider>, 'theme'>
) {
  return <ThemeProvider {...props} theme={viaTheme} />;
}

ThemeComponent.propTypes = {
  children: PropTypes.node
};

const profilerRender = (id: string, phase: string, actualDuration: number) => {
  if (actualDuration < 100) return;
  console.warn('Render took too long!', { id, phase, actualDuration });
};

export function renderWithTheme(component: React.ReactElement): RenderResult {
  const { ...viewUtils } = render(<ThemeComponent>{component}</ThemeComponent>);
  return {
    ...viewUtils,
    rerender: (rerenderComponent: React.ReactNode) =>
      viewUtils.rerender(
        <Profiler id="ThemeRender" onRender={profilerRender}>
          <ThemeComponent>{rerenderComponent}</ThemeComponent>
        </Profiler>
      )
  };
}

export const setClientWidth = (width: number) => {
  Object.defineProperty(document.documentElement, 'clientWidth', {
    writable: true,
    configurable: true,
    value: width
  });
};

export const setClientHeight = (height: number) => {
  Object.defineProperty(document.documentElement, 'clientHeight', {
    writable: true,
    configurable: true,
    value: height
  });
};
