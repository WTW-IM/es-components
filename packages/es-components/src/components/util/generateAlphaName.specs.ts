import generateAlphaName from './generateAlphaName';

describe('generateAlphaName', () => {
  it('generates a 6 character string', () => {
    const result = generateAlphaName();

    expect(result.length).toBe(6);
  });

  it('generates only alpha characters', () => {
    const result = generateAlphaName();

    expect(/^[a-zA-Z]+$/.test(result)).toBe(true);
  });

  it('generates a unique string each subsequent execution', () => {
    const result1 = generateAlphaName();
    const result2 = generateAlphaName();

    expect(result1).not.toEqual(result2);
  });
});
