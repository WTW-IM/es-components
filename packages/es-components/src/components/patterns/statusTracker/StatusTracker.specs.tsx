import React from 'react';
import StatusTracker from './StatusTracker';
import { renderWithTheme } from '../../util/test-utils';

describe('StatusTracker', () => {
  const planStates = ['step 1', 'step 2', 'step 3', 'step 4'];

  it('should display the third step with the active class', () => {
    const { container } = renderWithTheme(
      <StatusTracker statusArray={planStates} step={3} />
    );

    expect(container).toMatchSnapshot();
  });

  it('should display the first step with the active class', () => {
    const { container } = renderWithTheme(
      <StatusTracker statusArray={planStates} step={1} />
    );

    expect(container).toMatchSnapshot();
  });

  it('should display the last step with the active class', () => {
    const { container } = renderWithTheme(
      <StatusTracker statusArray={planStates} step={4} />
    );

    expect(container).toMatchSnapshot();
  });

  it('should display the last step with the error class when isErrorState is true', () => {
    const { container } = renderWithTheme(
      <StatusTracker isErrorState statusArray={planStates} step={4} />
    );

    expect(container).toMatchSnapshot();
  });
});
