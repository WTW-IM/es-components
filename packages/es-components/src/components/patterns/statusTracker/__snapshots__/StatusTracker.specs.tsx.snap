// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusTracker should display the first step with the active class 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 0;
  padding: 0;
  position: relative;
  z-index: 1;
}

.c1 {
  color: #919191;
  list-style-type: none;
  position: relative;
  text-align: center;
  width: 25%;
}

.c1 .status-text {
  display: none;
}

.c1.done > .status-text {
  color: #00a0d2;
}

.c1.active > .status-text,
.c1.error > .status-text {
  color: #00a0d2;
  display: inherit;
  font-weight: bold;
}

.c1:before {
  background-color: white;
  border: 3px solid #c5c5c5;
  border-radius: 50%;
  content: '';
  display: block;
  height: 20px;
  line-height: 20px;
  margin: 5px auto 15px auto;
  text-align: center;
  width: 20px;
}

.c1:after {
  background-color: #c5c5c5;
  content: '';
  height: 4px;
  left: -50%;
  position: absolute;
  top: 17px;
  width: 100%;
  z-index: -1;
}

.c1:first-child:after {
  content: none;
}

.c1.done:before {
  border-color: #00a0d2;
  color: #00a0d2;
  content: '\\e67b';
  font-family: 'bds-func-icons';
  font-size: 14px;
  font-weight: bold;
}

.c1.active:before {
  border-color: #00a0d2;
  color: #00a0d2;
  content: '\\e67b';
  font-family: 'bds-func-icons';
  font-size: 22px;
  font-weight: bold;
  height: 30px;
  line-height: 30px;
  margin: 0 auto 10px auto;
  width: 30px;
}

.c1.error:before {
  border-color: #cc0000;
  color: #cc0000;
  content: '\\e68e';
  font-family: 'bds-func-icons';
  font-size: 18px;
  font-weight: bold;
  height: 30px;
  line-height: 30px;
  margin: 0 auto 10px auto;
  width: 30px;
}

.c1.active:after,
.c1.done:after,
.c1.error:after {
  background-color: #00a0d2;
}

@media (min-width:768px) {
  .c1 .status-text {
    display: inherit;
  }

  .c1.active:before,
  .c1.done:before {
    content: '\\e67b';
    font-family: 'bds-func-icons';
    font-size: 22px;
    height: 30px;
    line-height: 30px;
    margin: 0 auto 10px auto;
    width: 30px;
  }
}

<div>
  <ol
    class="c0"
  >
    <li
      class="c1 active"
    >
      <span
        class="status-text"
      >
        step 1
      </span>
    </li>
    <li
      class="c1"
    >
      <span
        class="status-text"
      >
        step 2
      </span>
    </li>
    <li
      class="c1"
    >
      <span
        class="status-text"
      >
        step 3
      </span>
    </li>
    <li
      class="c1"
    >
      <span
        class="status-text"
      >
        step 4
      </span>
    </li>
  </ol>
</div>
`;

exports[`StatusTracker should display the last step with the active class 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 0;
  padding: 0;
  position: relative;
  z-index: 1;
}

.c1 {
  color: #919191;
  list-style-type: none;
  position: relative;
  text-align: center;
  width: 25%;
}

.c1 .status-text {
  display: none;
}

.c1.done > .status-text {
  color: #00a0d2;
}

.c1.active > .status-text,
.c1.error > .status-text {
  color: #00a0d2;
  display: inherit;
  font-weight: bold;
}

.c1:before {
  background-color: white;
  border: 3px solid #c5c5c5;
  border-radius: 50%;
  content: '';
  display: block;
  height: 20px;
  line-height: 20px;
  margin: 5px auto 15px auto;
  text-align: center;
  width: 20px;
}

.c1:after {
  background-color: #c5c5c5;
  content: '';
  height: 4px;
  left: -50%;
  position: absolute;
  top: 17px;
  width: 100%;
  z-index: -1;
}

.c1:first-child:after {
  content: none;
}

.c1.done:before {
  border-color: #00a0d2;
  color: #00a0d2;
  content: '\\e67b';
  font-family: 'bds-func-icons';
  font-size: 14px;
  font-weight: bold;
}

.c1.active:before {
  border-color: #00a0d2;
  color: #00a0d2;
  content: '\\e67b';
  font-family: 'bds-func-icons';
  font-size: 22px;
  font-weight: bold;
  height: 30px;
  line-height: 30px;
  margin: 0 auto 10px auto;
  width: 30px;
}

.c1.error:before {
  border-color: #cc0000;
  color: #cc0000;
  content: '\\e68e';
  font-family: 'bds-func-icons';
  font-size: 18px;
  font-weight: bold;
  height: 30px;
  line-height: 30px;
  margin: 0 auto 10px auto;
  width: 30px;
}

.c1.active:after,
.c1.done:after,
.c1.error:after {
  background-color: #00a0d2;
}

@media (min-width:768px) {
  .c1 .status-text {
    display: inherit;
  }

  .c1.active:before,
  .c1.done:before {
    content: '\\e67b';
    font-family: 'bds-func-icons';
    font-size: 22px;
    height: 30px;
    line-height: 30px;
    margin: 0 auto 10px auto;
    width: 30px;
  }
}

<div>
  <ol
    class="c0"
  >
    <li
      class="c1 done"
    >
      <span
        class="status-text"
      >
        step 1
      </span>
    </li>
    <li
      class="c1 done"
    >
      <span
        class="status-text"
      >
        step 2
      </span>
    </li>
    <li
      class="c1 done"
    >
      <span
        class="status-text"
      >
        step 3
      </span>
    </li>
    <li
      class="c1 active"
    >
      <span
        class="status-text"
      >
        step 4
      </span>
    </li>
  </ol>
</div>
`;

exports[`StatusTracker should display the last step with the error class when isErrorState is true 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 0;
  padding: 0;
  position: relative;
  z-index: 1;
}

.c1 {
  color: #919191;
  list-style-type: none;
  position: relative;
  text-align: center;
  width: 25%;
}

.c1 .status-text {
  display: none;
}

.c1.done > .status-text {
  color: #00a0d2;
}

.c1.active > .status-text,
.c1.error > .status-text {
  color: #00a0d2;
  display: inherit;
  font-weight: bold;
}

.c1:before {
  background-color: white;
  border: 3px solid #c5c5c5;
  border-radius: 50%;
  content: '';
  display: block;
  height: 20px;
  line-height: 20px;
  margin: 5px auto 15px auto;
  text-align: center;
  width: 20px;
}

.c1:after {
  background-color: #c5c5c5;
  content: '';
  height: 4px;
  left: -50%;
  position: absolute;
  top: 17px;
  width: 100%;
  z-index: -1;
}

.c1:first-child:after {
  content: none;
}

.c1.done:before {
  border-color: #00a0d2;
  color: #00a0d2;
  content: '\\e67b';
  font-family: 'bds-func-icons';
  font-size: 14px;
  font-weight: bold;
}

.c1.active:before {
  border-color: #00a0d2;
  color: #00a0d2;
  content: '\\e67b';
  font-family: 'bds-func-icons';
  font-size: 22px;
  font-weight: bold;
  height: 30px;
  line-height: 30px;
  margin: 0 auto 10px auto;
  width: 30px;
}

.c1.error:before {
  border-color: #cc0000;
  color: #cc0000;
  content: '\\e68e';
  font-family: 'bds-func-icons';
  font-size: 18px;
  font-weight: bold;
  height: 30px;
  line-height: 30px;
  margin: 0 auto 10px auto;
  width: 30px;
}

.c1.active:after,
.c1.done:after,
.c1.error:after {
  background-color: #00a0d2;
}

@media (min-width:768px) {
  .c1 .status-text {
    display: inherit;
  }

  .c1.active:before,
  .c1.done:before {
    content: '\\e67b';
    font-family: 'bds-func-icons';
    font-size: 22px;
    height: 30px;
    line-height: 30px;
    margin: 0 auto 10px auto;
    width: 30px;
  }
}

<div>
  <ol
    class="c0"
  >
    <li
      class="c1 done"
    >
      <span
        class="status-text"
      >
        step 1
      </span>
    </li>
    <li
      class="c1 done"
    >
      <span
        class="status-text"
      >
        step 2
      </span>
    </li>
    <li
      class="c1 done"
    >
      <span
        class="status-text"
      >
        step 3
      </span>
    </li>
    <li
      class="c1 error"
    >
      <span
        class="status-text"
      >
        step 4
      </span>
    </li>
  </ol>
</div>
`;

exports[`StatusTracker should display the third step with the active class 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 0;
  padding: 0;
  position: relative;
  z-index: 1;
}

.c1 {
  color: #919191;
  list-style-type: none;
  position: relative;
  text-align: center;
  width: 25%;
}

.c1 .status-text {
  display: none;
}

.c1.done > .status-text {
  color: #00a0d2;
}

.c1.active > .status-text,
.c1.error > .status-text {
  color: #00a0d2;
  display: inherit;
  font-weight: bold;
}

.c1:before {
  background-color: white;
  border: 3px solid #c5c5c5;
  border-radius: 50%;
  content: '';
  display: block;
  height: 20px;
  line-height: 20px;
  margin: 5px auto 15px auto;
  text-align: center;
  width: 20px;
}

.c1:after {
  background-color: #c5c5c5;
  content: '';
  height: 4px;
  left: -50%;
  position: absolute;
  top: 17px;
  width: 100%;
  z-index: -1;
}

.c1:first-child:after {
  content: none;
}

.c1.done:before {
  border-color: #00a0d2;
  color: #00a0d2;
  content: '\\e67b';
  font-family: 'bds-func-icons';
  font-size: 14px;
  font-weight: bold;
}

.c1.active:before {
  border-color: #00a0d2;
  color: #00a0d2;
  content: '\\e67b';
  font-family: 'bds-func-icons';
  font-size: 22px;
  font-weight: bold;
  height: 30px;
  line-height: 30px;
  margin: 0 auto 10px auto;
  width: 30px;
}

.c1.error:before {
  border-color: #cc0000;
  color: #cc0000;
  content: '\\e68e';
  font-family: 'bds-func-icons';
  font-size: 18px;
  font-weight: bold;
  height: 30px;
  line-height: 30px;
  margin: 0 auto 10px auto;
  width: 30px;
}

.c1.active:after,
.c1.done:after,
.c1.error:after {
  background-color: #00a0d2;
}

@media (min-width:768px) {
  .c1 .status-text {
    display: inherit;
  }

  .c1.active:before,
  .c1.done:before {
    content: '\\e67b';
    font-family: 'bds-func-icons';
    font-size: 22px;
    height: 30px;
    line-height: 30px;
    margin: 0 auto 10px auto;
    width: 30px;
  }
}

<div>
  <ol
    class="c0"
  >
    <li
      class="c1 done"
    >
      <span
        class="status-text"
      >
        step 1
      </span>
    </li>
    <li
      class="c1 done"
    >
      <span
        class="status-text"
      >
        step 2
      </span>
    </li>
    <li
      class="c1 active"
    >
      <span
        class="status-text"
      >
        step 3
      </span>
    </li>
    <li
      class="c1"
    >
      <span
        class="status-text"
      >
        step 4
      </span>
    </li>
  </ol>
</div>
`;
